pkg/pfcpsim/pfcpsim.go

```diff
@@ -101,60 +154,85 @@ func (c *PFCPClient) sendMsg(msg message.Message) error {
 		return err
 	}
 
-	if _, err := c.conn.Write(b); err != nil {
+	raddr, err := net.ResolveUDPAddr("udp", c.remoteAddr)
+	if err != nil {
+		return err
+	}
+
+	if _, err := c.conn.WriteTo(b, raddr); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (c *PFCPClient) receiveFromN4() {
-	buf := make([]byte, 1500)
+func (c *PFCPClient) receiveFromN4(ctx context.Context) {
+	buf := make([]byte, 3000)
 
 	for {
-		n, _, err := c.conn.ReadFrom(buf)
-		if err != nil {
-			continue
-		}
-
-		msg, err := message.Parse(buf[:n])
-		if err != nil {
-			continue
-		}
+		select {
+		case <-ctx.Done():
+			if c.cancelHeartbeats != nil {
+				c.cancelHeartbeats()
+			}
 
-		switch msg := msg.(type) {
-		case *message.HeartbeatResponse:
-			c.heartbeatsChan <- msg
+			err := c.conn.Close()
+			if err != nil {
+				fmt.Println(err)
+			}
 
-		case *message.SessionReportRequest:
-			// Ignore message
+			return
 		default:
-			c.recvChan <- msg
+			n, _, err := c.conn.ReadFrom(buf)
+			if err != nil {
+				continue
+			}
+
+			msg, err := message.Parse(buf[:n])
+			if err != nil {
+				continue
+			}
+
+			switch msg := msg.(type) {
+			case *message.HeartbeatResponse:
+				c.heartbeatsChan <- msg
+			case *message.HeartbeatRequest:
+				// ignore HeartbeatRequest
+				continue
+			case *message.SessionReportRequest:
+				if c.handleSessionReportRequest(msg) {
+					continue
+				}
+			default:
+				c.recvChan <- msg
+			}
 		}
 	}
 }
 
-func (c *PFCPClient) ConnectN4(remoteAddr string) error {
+func (c *PFCPClient) ConnectN4(ctx context.Context, remoteAddr string) error {
```

Comment:
@ianchen0119, why do you need to add context to this function? Does not it already have context when the `NewPFCPClient` is created/called (https://github.com/omec-project/pfcpsim/blob/main/pkg/pfcpsim/pfcpsim.go#L112)? I think this is related to the issue I am seeing with the Aether UPF with this latest `pfcpsim` change

Reply:
Hi @gab-arrobo @andy89923 

I Just created the PR to fix it: https://github.com/omec-project/pfcpsim/pull/74